utils/src/bitvec.rs

```diff
@@ -0,0 +1,971 @@
+                write!(f, "{}", if self.get_bit_unchecked(i) { "1" } else { "0" })?;
+            }
+        } else {
+            // Show first and last bits with ellipsis
+            for i in 0..HALF_DISPLAY {
+                write!(f, "{}", if self.get_bit_unchecked(i) { "1" } else { "0" })?;
+            }
+
+            write!(f, "...")?;
+
+            for i in (self.num_bits - HALF_DISPLAY)..self.num_bits {
+                write!(f, "{}", if self.get_bit_unchecked(i) { "1" } else { "0" })?;
+            }
+        }
+
+        write!(f, "]")
+    }
+}
+
+// ---------- Operations ----------
+
+impl Index<usize> for BitVec {
+    type Output = bool;
+
+    /// Allows accessing bits using the `[]` operator.
+    ///
+    /// Panics if out of bounds.
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        self.assert_index(index);
+        let value = self.get_bit_unchecked(index);
+        if value {
+            &true
+        } else {
+            &false
+        }
+    }
+}
+
+impl BitAnd for &BitVec {
+    type Output = BitVec;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        self.assert_eq_len(rhs);
+        let mut result = self.clone();
+        result.and(rhs);
+        result
+    }
+}
+
+impl BitOr for &BitVec {
+    type Output = BitVec;
+
+    fn bitor(self, rhs: Self) -> Self::Output {
+        self.assert_eq_len(rhs);
+        let mut result = self.clone();
+        result.or(rhs);
+        result
+    }
+}
+
+impl BitXor for &BitVec {
+    type Output = BitVec;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        self.assert_eq_len(rhs);
+        let mut result = self.clone();
+        result.xor(rhs);
+        result
+    }
+}
+
+// ---------- Codec ----------
+
+impl Write for BitVec {
+    fn write(&self, buf: &mut impl BufMut) {
+        varint::write(self.num_bits, buf);
+
+        // Write full blocks
+        for &block in &self.storage {
+            block.write(buf);
+        }
+    }
+}
+
+impl<R: RangeConfig> Read<R> for BitVec {
+    fn read_cfg(buf: &mut impl Buf, range: &R) -> Result<Self, CodecError> {
+        // Parse length
+        let num_bits = varint::read::<u32>(buf)? as usize;
+        if !range.contains(&num_bits) {
+            return Err(CodecError::InvalidLength(num_bits));
+        }
+
+        // Parse blocks
+        let num_blocks = num_bits.div_ceil(BITS_PER_BLOCK);
+        let mut storage = Vec::with_capacity(num_blocks);
+        for _ in 0..num_blocks {
+            let block = u64::read(buf)?;
+            storage.push(block);
+        }
+
+        // Ensure there were no trailing bits
+        let mut result = BitVec { storage, num_bits };
+        if result.clear_trailing_bits() {
+            return Err(CodecError::Invalid("BitVec", "trailing bits"));
+        }
+
+        Ok(result)
+    }
+}
+
+impl EncodeSize for BitVec {
+    fn encode_size(&self) -> usize {
+        varint::size(self.num_bits) + (u64::SIZE * self.storage.len())
+    }
+}
```

Comment:
Currently I just write all backing `u64` to the encoding. To save between 0 and 7 bytes, I could instead truncate the last block to the minimum number of required bytes, or encode it as a varint if it's not full.